### PR TITLE
  SONARPY-898 Avoid failing on older SonarLint 

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**

### CODE REVIEW

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java [hunk 1]
```diff
@@ -77,22 +77,18 @@
...
-
-  public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory, NoSonarFilter noSonarFilter,
-                      PythonCustomRuleRepository[] customRuleRepositories, AnalysisWarnings analysisWarnings) {
+                      PythonCustomRuleRepository[] customRuleRepositories, AnalysisWarningsWrapper analysisWarnings) {
     this(fileLinesContextFactory, checkFactory, noSonarFilter, customRuleRepositories, null, analysisWarnings);
   }
 
   public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory, NoSonarFilter noSonarFilter,
-                      @Nullable PythonCustomRuleRepository[] customRuleRepositories, @Nullable PythonIndexer indexer, @Nullable AnalysisWarnings analysisWarnings) {
+                      @Nullable PythonCustomRuleRepository[] customRuleRepositories, @Nullable PythonIndexer indexer, @Nullable AnalysisWarningsWrapper analysisWarnings) {
```
Comment (andrea-guarino-sonarsource) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java:91]: I believe  we can remove `@Nullable` annotation from `analysisWarning` parameter and on the corresponding field given that we always provide it.
Hence we might be able to remove the nullability check at line 117

**Review by andrea-guarino-sonarsource — COMMENTED** [review]
To me it's not clear what constructor will be called when `PythonIndexer` is present and `PythonCustomRuleRepository` is not.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java [hunk 2]
```diff
@@ -77,22 +77,18 @@
...
    * Constructor to be used by pico if neither PythonCustomRuleRepository nor PythonIndexer are to be found and injected.
    */
-  public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory, NoSonarFilter noSonarFilter, AnalysisWarnings analysisWarnings) {
-    this(fileLinesContextFactory, checkFactory, noSonarFilter, null, null, analysisWarnings);
+  public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory,
+    NoSonarFilter noSonarFilter, AnalysisWarningsWrapper defaultAnalysisWarningsWrapper) {
+    this(fileLinesContextFactory, checkFactory, noSonarFilter, null, null, defaultAnalysisWarningsWrapper);
   }
 
   public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory, NoSonarFilter noSonarFilter,
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java:84]: Thanks! Indeed, due to the missing constructor, `PythonIndexer` ended up not being injected properly...

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java [hunk 3]
```diff
@@ -88,14 +87,20 @@ public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactor
...
+  public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory, NoSonarFilter noSonarFilter,
+                      @Nullable PythonCustomRuleRepository[] customRuleRepositories, @Nullable PythonIndexer indexer, AnalysisWarningsWrapper analysisWarnings) {
     this.checks = new PythonChecks(checkFactory)
       .addChecks(CheckList.REPOSITORY_KEY, CheckList.getChecks())
       .addCustomChecks(customRuleRepositories);
     this.fileLinesContextFactory = fileLinesContextFactory;
     this.noSonarFilter = noSonarFilter;
     this.indexer = indexer;
     this.analysisWarnings = analysisWarnings;
+    LOG.info("CONSTRUCTOR CALLED: WARNING, RULE REPOSITORY AND INDEXER");
```
Comment (andrea-guarino-sonarsource) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java:103]: I think you forgot to remove this

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java [hunk 4]
```diff
@@ -77,29 +76,31 @@
...
-                      PythonIndexer indexer) {
-    this(fileLinesContextFactory, checkFactory, noSonarFilter, null, indexer, null);
+                      PythonCustomRuleRepository[] customRuleRepositories, AnalysisWarningsWrapper analysisWarnings) {
+    this(fileLinesContextFactory, checkFactory, noSonarFilter, customRuleRepositories, null, analysisWarnings);
   }
 
   public PythonSensor(FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory, NoSonarFilter noSonarFilter,
-                      PythonCustomRuleRepository[] customRuleRepositories, AnalysisWarnings analysisWarnings) {
-    this(fileLinesContextFactory, checkFactory, noSonarFilter, customRuleRepositories, null, analysisWarnings);
+    PythonIndexer indexer, AnalysisWarningsWrapper analysisWarnings) {
```
Comment (andrea-guarino-sonarsource) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java:90]: Let's be consistent about indentation; either we align with the parameters at the precedent line, or we uses two-space indentation everywhere (even on other constructors)

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonSensorTest.java [hunk 5]
```diff
@@ -319,7 +319,7 @@ public void cross_files_issues_only_one_file_analyzed() {
     // "mod.py" created but not added to context
     InputFile modFile = createInputFile("mod.py");
     PythonIndexer pythonIndexer = pythonIndexer(Arrays.asList(mainFile, modFile));
-    sensor(CUSTOM_RULES, pythonIndexer, null).execute(context);
+    sensor(null, pythonIndexer, analysisWarning).execute(context);
```
Comment (andrea-guarino-sonarsource) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonSensorTest.java:322]: why did you change CUSTOM_RULES -> null ?

Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonSensorTest.java:322]: I wanted to have a unit test with a `PythonIndexer` and without a `PythonCustomRuleRepository` to reflect the new constructor.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java (hunk 3 above)
Comment (guillaume-dequenne-sonarsource) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSensor.java]: Woops...Thanks! 😅

**Review by andrea-guarino-sonarsource — APPROVED** [review]
LGTM
